docs/protocol-methods.rst

```diff
@@ -293,57 +300,115 @@ blockchain.scripthash.get_history
 Return the confirmed and unconfirmed history of a :ref:`script hash
 <script hashes>`.
 
+A block interval ``[from_height, to_height[`` can be specified both by the
+client in the request if they do not want the whole history, and by the server
+in the response if the history was too large to fit.
+In the latter case, the client can tell they only received partial data
+by inspecting the *from_height* and *to_height* fields of the response,
+and initiate further requests for the missing interval.
+
 **Signature**
 
-  .. function:: blockchain.scripthash.get_history(scripthash)
+  .. function:: blockchain.scripthash.get_history(scripthash, from_height=0, to_height=-1,
+                                                  client_statushash=None, client_height=None)
```

Comment:
It seems wrong to be putting the burden of figuring out where the client is (presumably the point of the client_statushash and client_height) on the server.  I would have preferred to see something where client state doesn't enter the protocol.

Reply:
Right. This is the best I could come up with though. There are too many constraints when trying to support long histories...
Still, I think it is possible to implement this server-side in a way where it is relatively cheap to figure out the state of the client using these fields for well-behaving clients. By properly crediting resource usage cost to sessions, malicious or buggy clients can be banned.